### DISCUSSION TITLE
[v4] NextJS font variable not applying inside tailwind

### DISCUSSION THREAD

**Discussion by chandan-jal-evolution** [tailwindlabs/tailwindcss#13410] [Help] [answered]
With the new alpha version of tailwind I'm not being able to export the font family that is coming from the `NextJS`

Tailwind: 4.0.0-alpha.11
NextJs: 14.1.4

[Source Code](https://github.com/chandan-jal-evolution/next-tailwindcss-alpha-test)

**Comment by adamwathan** [comment, marked as answer]
Hey! Don't forget to follow the instructions in the Next docs for setting those font variables:

https://nextjs.org/docs/pages/building-your-application/optimizing/fonts#with-tailwind-css

Here's a working version of your `layout.tsx` file:

```jsx
import type { Metadata } from "next";
import { Montserrat, Poppins, PT_Serif } from "next/font/google";
import "./globals.css";

const montserrat = Montserrat({
  subsets: ["latin"],
  variable: "--font-montserrat",
});

const poppins = Poppins({
  subsets: ["latin"],
  weight: ["500"],
  variable: "--font-poppins",
});

const ptSerif = PT_Serif({
  subsets: ["latin"],
  weight: ["400"],
});

export const metadata: Metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body
        className={`${ptSerif.className} ${montserrat.variable} ${poppins.variable}`}
      >
        {children}
      </body>
    </html>
  );
}
```

Here's the important part:

```diff
    <html lang="en">
      <body
-       className={ptSerif.className}
+       className={`${ptSerif.className} ${montserrat.variable} ${poppins.variable}`}
      >
        {children}
      </body>
    </html>
```

**Reply by chandan-jal-evolution** [reply]
Hey @adamwathan, Thanks for your response.

**Reply by seedrix** [reply]
Did that work for you?
In the docs it says to use --font-sans and not --font-family-sans: https://tailwindcss.com/docs/theme#relationship-to-utility-classes. But still it does not seem to be working for me.

**Reply by wongjn** [reply]
You might have to do `@theme inline {` or add the class names to the `html` element.

**Reply by seedrix** [reply]
That works. Wow, thanks for the quick response!

**Reply by cnrad** [reply]
Changing `@theme {` to `@theme inline {` worked, thank you 🙏 

For future readers: this happens if you use the `variable` option for NextJS fonts and try to define the font variables in `@theme` by referencing the ones from NextJS - this is actually [explained pretty clearly in the docs](https://tailwindcss.com/docs/theme#referencing-other-variables) (albeit a bit buried)

**Reply by kylemh** [reply]
Hey folks

Still a lil confused...

I've got the following:
```postcss
@import 'tailwindcss';

:root {
  --background: #ffffff;
  --foreground: #171717;

  /* base theme variables */
}

@media (prefers-color-scheme: dark) {
  :root {
    --background: #0a0a0a;
    --foreground: #ededed;

    /* other dark theme overrides */
  }
}

@theme {
  /* big-ass theme */
}

@theme inline {
  --font-sans: var(--font-geist);
}

body {
  background: var(--background);
  color: var(--foreground);
}

```

My root layout file (`src/app/layout.tsx`) looks like this:

```tsx
import { Geist } from 'next/font/google';
import '@/globals.css';

const geist = Geist({
  variable: '--font-geist',
  subsets: ['latin'],
  fallback: ['system-ui', 'Helvetica', 'Arial', 'sans-serif'],
});

export default async function RootLayout({
  children,
}: Readonly<{ children: React.ReactNode }>) {
  return (
    <html lang="en">
      <body className={`${geist.variable} antialiased`}>
        {children}
      </body>
    </html>
  );
}

```

I've tried adding `--font-sans: var(--font-geist)` to the root and I've also tried adding:
```css
html {
  font-family: var(--font-geist);
}
```
and
```css
html {
  font-family: var(--font-sans);
}
```

but both CSS variables show up as undefined in devtools

it looks like I need to define

```
body {
  font-family: var(--font-geist);
}
```

if i try `var(--font-sans)` the inline theme override doesn't appear to be working

**Reply by wongjn** [reply]
Try adding `geist.variable` to `<html>` instead:

```jsx
export default async function RootLayout({
  children,
}: Readonly<{ children: React.ReactNode }>) {
  return (
    <html lang="en" className={geist.variable}>
      <body className="antialiased">
        {children}
      </body>
    </html>
  );
}
```

**Reply by kylemh** [reply]
Okay, that works! I also see the source of my problem. I never looked at [these docs](https://nextjs.org/docs/app/building-your-application/optimizing/fonts#with-tailwind-css) because the `create-next-app` exec puts the className on the `<body>` element

https://github.com/vercel/next.js/pull/78055 tryna fix that issue so nobody else encounters it.

**Comment by gijsbotje** [comment]
Is there a way to fix this for the pages router as well?

I've tried adding a custom font-family to the config like this:
```
--font-heading: var(--my-font, sans-serif), -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Ubuntu, 'Helvetica Neue', sans-serif;
```
But the browsers says `--my-font` isn't defined. The var `--my-font` gets set to the first div in the `<div id="__next">`. 

I also tried just using the var as a custom value with `font-(--my-font)` as described in the docs [here](https://tailwindcss.com/docs/font-family#using-a-custom-value:~:text=For%20CSS%20variables%2C%20you%20can%20also%20use%20the%20font%2D(%3Ccustom%2Dproperty%3E)%20syntax%3A). This unfortunately thinks it's a font-weight custom value.

**Reply by gijsbotje** [reply]
Just found a solution for this. I saw the portion in the next docs about setting a font-family in a style tag in the head ([link](https://nextjs.org/docs/pages/building-your-application/optimizing/fonts#apply-the-font-in-head)). I used that example to create the var in a style tag in the head so tailwind's `:root` defined vars can read it.
```jsx
<style jsx global>{`
  :root {
    --my-font: ${inter.style.fontFamily};
  }
`}</style>
```

**Reply by dzuloaga** [reply]
Thanks! If you using the App router you can use plain style:

````css
<style>
  {`
  :root {
    --font-inter: ${inter.style.fontFamily};
    --font-playfair-display: ${playfairDisplay.style.fontFamily};
  }
`}
</style>```
````
